File: python/src/feb/week2/boj/문자열게임2_박재환.py
def solution():
    import sys
    input = sys.stdin.readline

    tc = int(input())
    for _ in range(tc):
        word = input().strip()
        k = int(input())

        alpha = [[] for _ in range(26)]

        for i in range(len(word)):
            id = ord(word[i]) - ord('a')
            alpha[id].append(i)

        max_len = -1
        min_len = len(word)+1

        for i in range(26):
            for j in range(len(alpha[i])-k+1):
                s = alpha[i][j]
                e = alpha[i][j+k-1]
                l = e-s+1
                min_len = min(min_len, l)
                max_len = max(max_len, l)
        if min_len == len(word)+1 or max_len == -1:
            print(-1)
        else:
            print(min_len,max_len)

if __name__ == '__main__':
    solution()

File: java/src/feb/week2/boj/문자열게임2_박재환.java
package feb.week2.boj;

import java.util.*;
import java.io.*;

public class 문자열게임2_박재환 {
    static BufferedReader br;
    static StringBuilder sb;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        sb = new StringBuilder();
        init();
        br.close();
        System.out.println(sb);
    }
    static void init() throws IOException {
        int tc = Integer.parseInt(br.readLine().trim());
        while(tc-- > 0) {
            String word = br.readLine().trim();
            int target = Integer.parseInt(br.readLine().trim());

            int minLen = word.length()+1;
            int maxLen = -1;

            List<Integer>[] alpha = new List[26];
            for(int i=0; i<26; i++) alpha[i] = new ArrayList<>();

            for(int i=0; i<word.length(); i++) {
                char c = word.charAt(i);
                alpha[c-'a'].add(i);            // 해당 문자의 인덱스 위치 기억
            }

            /**
             * 각 알파벳을 확인하며 구간 갱신
             */
            for(List<Integer> list : alpha) {
                for(int i=0; i<=list.size()-target; i++) {
                    int s = list.get(i);
                    int e = list.get(i+target-1);
                    int len = e-s+1;

                    minLen = Math.min(minLen, len);
                    maxLen = Math.max(maxLen, len);
                }
            }

            if(minLen == word.length()-1 || maxLen == -1) sb.append(-1);
            else sb.append(minLen).append(' ').append(maxLen);
            sb.append('\n');
        }
    }
}
